staging/src/k8s.io/apimachinery/pkg/watch/mux.go

```diff
@@ -209,16 +210,30 @@ func (m *Broadcaster) closeAll() {
 	m.watchers = map[int64]*broadcasterWatcher{}
 }
 
-// Action distributes the given event among all watchers.
+// Action distributes the given event among all watchers. It panics if Shutdown()
+// has already been called.
 func (m *Broadcaster) Action(action EventType, obj runtime.Object) {
 	m.incoming <- Event{action, obj}
 }
 
 // Action distributes the given event among all watchers, or drops it on the floor
-// if too many incoming actions are queued up.  Returns true if the action was sent,
-// false if dropped.
+// if too many incoming actions are queued up or if Shutdown() has already been called.
+// Returns true if the action was sent, false if dropped.
 func (m *Broadcaster) ActionOrDrop(action EventType, obj runtime.Object) bool {
+	// select ordering is not deterministic, and it would match adding the event to
+	// the incoming channel. A separate select is needed to avoid this problem.
+	// https://golang.org/ref/spec#Select_statements
 	select {
+	case <-m.stopped:
+		return false
+	default:
+	}
+
+	select {
+	// very unlikely that the broadcaster might be stopped while this is called,
+	// but just to be on the safe side.
```

Comment:
As you mentioned earlier,  the order of the select is not deterministic. So if the situation in the comment happens, this can still panic, if `case m.incoming <- Event{action, obj}` is selected.

Reply:
I'm not 100% sure that it's even possible to begin with. I would guess that the compiler would optimize for that and would combine both select statements and chose the first with preferences, but it's just a guess.

Reply:
It can still panic. Try out this [example](https://play.golang.org/p/iSq4ian78B2), sometimes it panics, sometimes it prints out "stopped".

I have two thoughts:
* Can we just leave the `incoming` channel open? As long as no one references it, its memory will be garbage collected.
* Can we just do a Recover() in both Action and ActionOrDrop?

Reply:
Seems if we follow the pattern proposed in this PR, it doesn't panic at least in playground https://play.golang.org/p/UQY9nPy7jK8

Reply:
> Seems if we follow the pattern proposed in this PR, it doesn't panic at least in playground https://play.golang.org/p/UQY9nPy7jK8

@caesarxuchao is correct, this is still not correct, it just narrows the window where a panic can happen. Serializing an async call to Shutdown into the playground, it looks like this: https://play.golang.org/p/gC8BuU_7f8p